node/src/components/chain_synchronizer/operations.rs

```diff
@@ -341,125 +347,196 @@ impl CanUseSyncingNodes for Deploy {}
 impl CanUseSyncingNodes for BlockAndDeploys {}
 impl CanUseSyncingNodes for BlockHeadersBatch {}
 
-/// Returns fully-connected, non-syncing peers that are known to be not banned.
-async fn get_filtered_fully_connected_non_syncing_peers<REv>(
-    ctx: &ChainSyncContext<'_, REv>,
-) -> Vec<NodeId>
+/// Gets a list of peers suitable for the fetch operation.
+async fn get_peers<REv>(include_syncing: bool, ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
 where
     REv: From<NetworkInfoRequest>,
 {
-    let mut peer_list = ctx
-        .effect_builder
-        .get_fully_connected_non_syncing_peers()
-        .await;
+    let mut peer_list = if include_syncing {
+        ctx.effect_builder.get_fully_connected_peers().await
+    } else {
+        ctx.effect_builder
+            .get_fully_connected_non_syncing_peers()
+            .await
+    };
     ctx.filter_bad_peers(&mut peer_list);
     peer_list
 }
 
-/// Returns fully-connected, syncing and non-syncing peers that are known to be not banned.
-async fn get_filtered_fully_connected_peers<REv>(ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
+/// Fetching should stop only when the network is connected and all retries have been exhausted.
+fn should_stop_fetching(
+    retry_count: &mut usize,
+    has_connected_to_network: bool,
+    max_sync_fetch_attempts: usize,
+) -> bool {
+    if !has_connected_to_network {
+        return false;
+    }
+
+    *retry_count += 1;
+    if *retry_count <= max_sync_fetch_attempts {
+        return false;
+    }
+
+    true
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum FetchWithRetryError<T>
 where
-    REv: From<NetworkInfoRequest>,
+    T: Item,
 {
-    let mut peer_list = ctx.effect_builder.get_fully_connected_peers().await;
-    ctx.filter_bad_peers(&mut peer_list);
-    peer_list
+    #[error("Fetch retries exhausted for item with id {id:?}. Total attempts: {total_attempts}, retries while network was connected: {retries_while_connected}")]
+    RetriesExhausted {
+        id: T::Id,
+        total_attempts: usize,
+        retries_while_connected: usize,
```

Comment:
I'd rename this to something like `attempts_after_bootstrapped`.  I think using the term "retries" in the context of counters is confusing (do you count the first fetch attempt as a retry or not?).  So generally in this PR I'd be inclined to swap out the term "retry" for "attempt" where it's used to represent a count.

Reply:
Renamed in 6fe8758fa5d45fa61a77bc4ecc38b76cead2b880